chat/src/main/java/io/skygear/plugins/chat/ui/ConversationFragment.kt

```diff
@@ -445,32 +458,73 @@ open class ConversationFragment() :
             val message = ChatMessage()
             message.asset = asset
             message.metadata = meta
+            this.fragmentMessageSentListener?.onBeforeMessageSent(this, message)
             addMessage(message, Uri.parse("file://" + voiceRecordingFileName))
-
             this.skygearChat?.addMessage(message, conv, object : SaveMessageCallback {
-                override fun onSucc(chatMsg: ChatMessage?) {
+                override fun onSucc(message: ChatMessage?) {
                     voiceRecordingFile.delete()
+                    callMessageSentSuccessListener(message)
                 }
 
                 override fun onSaveResultCached(message: io.skygear.plugins.chat.Message?) {
+                    callMessageSentSuccessListener(message)
                 }
 
                 override fun onFail(error: Error) {
                     Log.e(
                             ConversationFragment.TAG,
                             "Failed to send voice message: ${error.message}"
                     )
+                    this@ConversationFragment.fragmentMessageSentListener?.onMessageSentFailed(this@ConversationFragment, message, error)
                 }
             })
         }
     }
 
+    fun callMessageSentSuccessListener(chatMsg: ChatMessage?) {
+        chatMsg?.let {
+            this@ConversationFragment.fragmentMessageSentListener?.
+                    onMessageSentSuccess(this@ConversationFragment, chatMsg)
+        } ?: run {
+            this@ConversationFragment.fragmentMessageSentListener?.
+                    onMessageSentFailed(this@ConversationFragment,
+                            null, Error("No message sent"))
+        }
+    }
+
+
+    fun callMessageFetchSuccessListener(chatMsgs: List<ChatMessage> ?) {
+        chatMsgs?.let {
+            this@ConversationFragment.fragmentMessageFetchListener?.onMessageFetchSuccess(this@ConversationFragment, it)
+        } ?: run {
+            this@ConversationFragment.fragmentMessageFetchListener?.
+                    onMessageFetchFailed(this@ConversationFragment,
+                            Error("No message fetched"))
```

Comment:
I think it should not be an error if there is no message? We will have this case when user enter a new conversation without message, or when user scroll to the top and reach the top.